### PR TITLE
Windows: Emulate MSG_PEEK by read

### PR DESCRIPTION
**Pull request by soulxu**
Implementing peek by draining the socket and storing the
content of the socket to a buffer.

Additional Description: See #17395
Risk Level: Medium
Testing: Automated
Docs Changes: Pending
Release Notes: Pending
Platform-Specific Features: Windows only

Signed-off-by: He Jie Xu <hejie.xu@intel.com>
Co-authored-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

### CODE REVIEW

**Comment by soulxu** [comment]
/cc @adisuissa @alyssawilk 

thanks @adisuissa give me the error info, and apologize for not making the code right.

@adisuissa said there still other errors, appreciate if can help me spot them out, then let me fix it.

File: source/common/buffer/buffer_impl.cc [hunk 1]
```diff
@@ -118,6 +118,43 @@ void OwnedImpl::copyOut(size_t start, uint64_t size, void* data) const {
...
+  while (num_dest_slices_read < num_slice && num_bytes_read < total_length_to_read) {
+    const Slice& src_slice = slices_[num_src_slices_read];
+    const Buffer::RawSlice& dest_slice = dest_slices[num_dest_slices_read];
+    uint64_t left_to_read = total_length_to_read - num_bytes_read;
+    uint64_t left_data_size_in_dst_slice = dest_slice.len_ - dest_slice_offset;
+    uint64_t left_data_size_in_src_slice = src_slice.dataSize() - src_slice_offset;
+    // The length to copy should be size of smallest in the source slice available size and
+    // the dest slice available size.
+    uint64_t length_to_copy =
+        std::min(left_data_size_in_src_slice, std::min(left_data_size_in_dst_slice, left_to_read));
```
Comment (soulxu) [inline source/common/buffer/buffer_impl.cc:138]: I correct the `length_to_copy` calculation.

File: source/common/buffer/buffer_impl.cc [hunk 2]
```diff
@@ -118,6 +118,43 @@ void OwnedImpl::copyOut(size_t start, uint64_t size, void* data) const {
...
+    if (src_slice_offset == src_slice.dataSize()) {
+      num_src_slices_read++;
+      src_slice_offset = 0;
+    }
+    if (dest_slice_offset == dest_slice.len_) {
+      num_dest_slices_read++;
+      dest_slice_offset = 0;
+    }
+    ASSERT(src_slice_offset <= src_slice.dataSize());
+    ASSERT(dest_slice_offset <= dest_slice.len_);
```
Comment (soulxu) [inline source/common/buffer/buffer_impl.cc:152]: Add two assertions for defending.

**Comment by adisuissa** [comment]
> @adisuissa said there still other errors, appreciate if can help me spot them out, then let me fix it.

There is a fuzz error that seems to be in the fuzz test itself, and not due to this PR.
The error is being investigated separately.

**Comment by soulxu** [comment]
> > @adisuissa said there still other errors, appreciate if can help me spot them out, then let me fix it.
> 
> There is a fuzz error that seems to be in the fuzz test itself, and not due to this PR. The error is being investigated separately.

got it, thanks

**Comment by alyssawilk** [comment]
tagging Adi for first pass since @wrowe has not been responding.

**Comment by wrowe** [comment]
@davinci26 as much of this was your code, and we still have you on the assignable, envoy-triage and windows-dev teams, are you willing to take a pass at this? [If you would like us to modify your teams we can do that, too.]

**Comment by davinci26** [comment]
I will review later today

File: test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc [hunk 3]
```diff
@@ -889,6 +912,20 @@ TEST_P(ProxyProtocolTest, V2Fragmented5Error) {
   TestThreadsafeSingletonInjector<Api::OsSysCallsImpl> os_calls(&os_sys_calls);
 
   // TODO(davinci26): Mocking should not be used to provide real system calls.
+#ifdef WIN32
+  bool partial_write = false;
+  EXPECT_CALL(os_sys_calls, readv(_, _, _))
+      .Times(AnyNumber())
+      .WillRepeatedly(Invoke([&](os_fd_t fd, const iovec* iov, int num_iov) {
+        if (partial_write) {
+          ENVOY_LOG_MISC(debug, "inject failure");
```
Comment (davinci26) [inline test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc:921]: nit: do we want to keep those logs?

File: test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc [hunk 4]
```diff
@@ -889,6 +912,20 @@ TEST_P(ProxyProtocolTest, V2Fragmented5Error) {
...
+          return Api::SysCallSizeResult{-1, 0};
+        }
+        ENVOY_LOG_MISC(debug, "wire");
```
Comment (davinci26) [inline test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc:924]: ditto

**Review by davinci26 — APPROVED** [review]
overall lgtm, thanks for all the hard work on this

File: test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc (hunk 3 above)
Comment (soulxu) [reply test/extensions/filters/listener/proxy_protocol/proxy_protocol_test.cc:921]: yea, can be removed.

**Comment by soulxu** [comment]
> overall lgtm, thanks for all the hard work on this

thanks! done the clean up for those useless logs in the tests.

**Pull request merged**